azure-core/azure/core/pipeline/policies/retry.py

```diff
@@ -0,0 +1,251 @@
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class RetryPolicy(HTTPPolicy):
+
+    #: Maximum backoff time.
+    BACKOFF_MAX = 120
+
+    def __init__(self, **kwargs):
+        self.total_retries = kwargs.pop('retry_total', 10)
+        self.connect_retries = kwargs.pop('retry_connect', 3)
+        self.read_retries = kwargs.pop('retry_read', 3)
+        self.status_retries = kwargs.pop('retry_status', 3)
+        self.backoff_factor = kwargs.pop('retry_backoff_factor', 0.8)
+        self.backoff_max = kwargs.pop('retry_backoff_max', self.BACKOFF_MAX)
+
+        safe_codes = [i for i in range(500) if i != 408] + [501, 505]
+        retry_codes = [i for i in range(999) if i not in safe_codes]
+        status_codes = kwargs.pop('retry_on_status_codes', [])
+        self._retry_on_status_codes = set(status_codes + retry_codes)
+        self._method_whitelist = frozenset(['HEAD', 'GET', 'PUT', 'DELETE', 'OPTIONS', 'TRACE'])
+        self._respect_retry_after_header = True
+
+    @classmethod
+    def no_retries(cls):
+        return cls(retry_count_total=0)
+
+    def configure_retries(self, **kwargs):
+        return {
+            'total': kwargs.pop("retry_total", self.total_retries),
+            'connect': kwargs.pop("retry_connect", self.connect_retries),
+            'read': kwargs.pop("retry_read", self.read_retries),
+            'status': kwargs.pop("retry_status", self.status_retries),
+            'backoff': kwargs.pop("retry_backoff_factor", self.backoff_factor),
+            'max_backoff': kwargs.pop("retry_backoff_max", self.BACKOFF_MAX),
+            'methods': kwargs.pop("retry_on_methods", self._method_whitelist),
+            'history': []
+        }
+
+    def get_backoff_time(self, settings):
+        """ Formula for computing the current backoff
+
+        :rtype: float
+        """
+        # We want to consider only the last consecutive errors sequence (Ignore redirects).
+        consecutive_errors_len = len(settings['history'])
+        if consecutive_errors_len <= 1:
+            return 0
+
+        backoff_value = settings['backoff'] * (2 ** (consecutive_errors_len - 1))
+        return min(settings['max_backoff'], backoff_value)
+
+    def parse_retry_after(self, retry_after):
+        # Whitespace: https://tools.ietf.org/html/rfc7230#section-3.2.4
+        if re.match(r"^\s*[0-9]+\s*$", retry_after):
+            seconds = int(retry_after)
+        else:
+            retry_date_tuple = email.utils.parsedate(retry_after)
+            if retry_date_tuple is None:
+                raise InvalidHeader("Invalid Retry-After header: %s" % retry_after)
+            retry_date = time.mktime(retry_date_tuple)
+            seconds = retry_date - time.time()
+
+        if seconds < 0:
+            seconds = 0
+
+        return seconds
+
+    def get_retry_after(self, response):
+        """ Get the value of Retry-After in seconds. """
+
+        retry_after = response.http_response.headers.get("Retry-After")
```

Comment:
Do we want to honor x-ms-retry-after-ms in addition to Retry-After in the "default" retry policy?

Reply:
Hmm probably? By depending on a library called "azure-core" one has already opted-in to some MS-specific stuff and checking for an extra header wont otherwise affect the policy behaviour.